src/blockchain.rs

```diff
@@ -1,28 +1,47 @@
 use crate::block::{Block, BlockHeader};
 use crate::utxoset::UTXOSet;
+use std::collections::HashMap;
 
+/// We want to keep information on the block header, the SHA256hash,
+/// and whether or not the block is on the longest chain
 pub type BlockIndex = (BlockHeader, [u8; 32]);
 /// Indexes of chain attribute
 #[derive(Debug, Clone)]
 pub struct BlockchainIndex {
-    /// Vector of blocks
-    blocks: Vec<BlockIndex>,
+    /// Map of Blocks
+    block_map: HashMap<[u8; 32], BlockIndex>,
+    // HashMap of `Block`s that are on the longest chain
+    longest_chain_map: HashMap<[u8; 32], bool>,
+    /// The block has of the latest `Block` on the longest chain
+    last_block_hash: Option<[u8; 32]>,
+    /// The `Block` at the end of the epoch
+    genesis_block_hash: Option<[u8; 32]>,
```

Comment:
I was hoping to refactor block.rs before we do this work because it woudl be easier to do in that order... this is fine...

My intuition is that longest_chain should be a stack of block_hash(the height of the stack is the block_id). last_block_hash is then just a method that does longest_chain_stack.peek(). 

I'm still confused everytime i see the word "genesis" in a Saito context... What does genesis_block_hash represent? Will this get changed every time we do add_block?

We'll need a way to track the bottom of the epoch for rebroadcasting. Is this the purpose of genesis_block_hash, I think we again would need a data structure like block_id -> block_hash. The stack i mentioned above almost works, but it needs to also work like a queue. I.E. we need a to store block hashes and have both Queue(FIFO) and a Stack(LIFO) properties, this is a Double-ended queue. This seems correct to me because it's basically what an epoch looks like.

So, I think we need the block_map as is and everything else can be managed by a Deque:

epoch_block_hashes: Deque;

std seems to have a VecDeque, but we should do a bit more research before we choose which deque to use. https://doc.rust-lang.org/std/collections/struct.VecDeque.html